src/backends/simple.ts

```diff
@@ -0,0 +1,442 @@
+                const filteredResult = await this.filterer.filter([{ name: 'in0:v', frames: this.frames }]);
+                filteredFrames = filteredResult.flatMap(r => r.frames);
+                LOG && console.log('filteredFrames', filteredFrames.length, 'filteredFrames.pts:', filteredFrames.map(f => f.pts), '-> target.pts:', targetPTS);
+                this.filteredFrames = filteredFrames as beamcoder.DecodedFrames;
+
+                // get closest frame
+                const closestFrame = filteredFrames.reverse().find(f => f.pts <= targetPTS);
+                closestFramePTS = closestFrame?.pts;
+                LOG && console.log('closestFramePTS', closestFramePTS, 'targetPTS', targetPTS);
+                if (!outputFrame || closestFramePTS < targetPTS) {
+                    LOG && console.log('|--> assigning outputFrame', closestFrame?.pts);
+                    outputFrame = closestFrame;
+                }
+                else {
+                    // break out of the loop if we've found the closest frame (and ensure we don't move to the next
+                    // packet by calling _getNextPacketAndDecodeFrames again) as this risks us getting a frame that is
+                    // after our targetPTS
+                    LOG && console.log('breaking');
+                    break;
+                }
+            }
+            // get the next packet and frames
+            ({ packet: this.packet, frames: this.frames } = await this._getNextPacketAndDecodeFrames());
+
+            // keep track of how many packets we've read
+            if (packetReadCount >= MAX_PACKET_READS) {
+                throw Error('Reached max packet reads');
+            }
+            packetReadCount++;
+        }
+
+        if (!outputFrame) {
+            throw Error('No matching frame found');
+        }
+        LOG && console.log('read', packetReadCount, 'packets');
+
+        this.previousTargetPTS = targetPTS;
+        return outputFrame;
+    }
+
+    /**
+     * Get the next packet from the video stream and decode it into frames. Each frame has a presentation time stamp.
+     * If we've reached the end of the stream and no more packets are available, we'll extract the last frames from
+     * the decoder and destroy it.
+     */
+    async _getNextPacketAndDecodeFrames() {
+        const packet = await this._getNextVideoStreamPacket();
+        LOG && console.log('packet pts:', packet?.pts);
+
+        // NOTE: maybe we should only decode frames when we're relatively close to our targetPTS?
+        // extract frames from the packet
+        let decodedFrames = null;
+        if (packet !== null && this.decoder) {
+            decodedFrames = await this.decoder.decode(packet as Packet);
+            LOG && console.log('decodedFrames', decodedFrames.frames.length, decodedFrames.frames.map(f => f.pts));
+        }
+        // we've reached the end of the stream
+        else {
+            if (this.decoder) {
```

Comment:
This `if` is unnecessarily nested. Could be `else if` ... `else`

Reply:
And I think I wrote it lol

Reply:
That's true. I just prefer it this way is it shows the state for the other situation which is important to understand when stepping through the code. Maybe we should put this in a comment instead?

Reply:
Ah if it's clearer, we can leave as-is. That's a detail honestly.